README.md
# Simple Blackjack

## Project Description and Introduction
Simple Blackjack was created as an elementary version of the original Blackjack poker game. Simple Blackjack is played between the dealer and one player.

To play the game click [here.](https://bdfroehlich.github.io/MP1-SimpleBlackJack/?# "This link will take you to Simple Blackjack.")

### How to Use
Upon launching the game the player will be prompted to place a bet. Once a bet is placed and submitted the player may hit the deal button to start the hand. Once the first two cards are dealt to the player and the dealer, the player can choose to hit (take another card) or stand (keep their current cards and allow the dealer to draw if they can). Once the dealer has drawn their cards the outcome of the hand will be displayed to the screen and the players bet amount will be added or taken away from their total currency depending on if they win or lose. After the hand is finished the player can start a new round.

Blackjack rules **Ammended for Simple Blackjack** [here.](https://www.ildado.com/blackjack_rules.html "This link will take you the rules of Blackjack.")

### Technologies
- Bootstrap v5.1
- HTML
- CSS
- Javascript
- SCSS
- Typescript

### Setup
To run this project in your browser launch it via the published link in the repository or from the link above.

### CSS vs. SCSS
![original game with raw css](https://github.com/bdfroehlich/MP1-SimpleBlackJack/blob/main/Images/game_original.JPG?raw=true)
![game with scss](https://github.com/bdfroehlich/MP1-SimpleBlackJack/blob/main/Images/game.JPG?raw=true)

### Challenges
- The amount of comparative logic that went into checking the dealer and player hands when the turns ended.
- Adding weighted values for face cards until discovering the use of dictionaries and passing it into the Card class constructor.
- Correcting the logic for placeBet and updateCurrency so the players currency would not revert back to the default amount after each round.

### Future Features & Updates
- Make better use of class constructors and class methods.
- Implement other play features and bets including doubling down and splitting.
- Add logic for Aces to be equal to 1 or 11.
- Add media queries for different media types/devices and clean up styling.
- Sound files for theme music and a sound effect to be played when cards are dealt.
- Total currency needs to be a decimal.
- Add some CSS effects to make the cards being dealt appear to come off the center deck.

### Issues
- On initialization of the page the player bet is not accepted on the first submit. It must be submitted a second time. Betting will continue to work as intended after.

### Sources
PNGs of playing cards: [here.](https://code.google.com/archive/p/vector-playing-cards "This link will take you the deck of cards PNG download.")



CHANGELOG.md
- updating images src
- Merge pull request #1 from bdfroehlich/scss-styling
- adding images to readme
- updating readme
- removing bootstrap
- cleaning up style.scss and making code more DRY
- cleaning up styling, removing redundant styling
- cleaning up player detail div items
- additional mixins and style.scss cleanup
- cleaning up scss styling
- updating styling
- adding alertconfig partial file
- fixing index.html issue
- updating styling
- commit
- cleaning code
- updating readme
- updating readme
- cleaning up code
- updating readme
- commit
- re-organizing readme
- clean up code
- updating sources
- updating notes in createDeck
- adding source for creating dictionaries
- added source to shuffle function
- updating readme
- adding link to game in readme
- updating title on browser tab
- adding to future features
- I think I have finally resolved the getBet and updateCurrency issues..
- fixed issue with winning twice in a row - winnings needed to be set back to none once the table was cleared - fixed issue with hard hitting two aces
- corrected outcome logic for if the dealers hand was = 17 and updated submit/deal above text
- Merge branch 'main' of https://github.com/bdfroehlich/MP1-SimpleBlackJack
- updated intial place bet and deal text
- Set theme jekyll-theme-slate
- Set theme jekyll-theme-cayman
- updating md
- added href to back of card image
- updating md
- fixed currency calculation error not taking bet from player when dealer won
- fixed dealer hit function by removing conditional and adding it to checkdealer total, fixed issue with winnings and currency values in updateCurrency
- adding audio file updated fade styling
- experimenting with audio files and sound effects
- updating readme
- setTimeouts for initial deal
- cleaned up code
- updating readme
- fixed return values from updateCurrency()
- added fade-in & fade-out styling and source
- added function comments for readability
- updated functionality
- updating readme
- updating readme
- updated alert id styling
- added h4 element for various alerts
- fixed getBet, updateCurrency, and added currency checker to check for zero currency
- updating readme
- updating readme
- updating readme
- updating readme
- updating readme
- updating readme
- updating readme
- working on fixing bet/updatecurrency functions
- working on bet logic/value updates and added text for player to bet and click deal
- fixed some errors in outcome logic
- changed outcome element tag to h3
- added boolean so player can not immediately start new round after placing bet
- updated styles to match new form div and form IDs
- updated default bet value, revised form div and form classes to separate IDs so .reset could be used on form
- currency updates when bet submitted, form clears and bet amount clears when new round is started
- cleaned up form element, cleaned up styles
- updating bet function, added boolean for bet submission, cleaned up selectors
- working on betsubmit function
- updated form and styling
- removed bet buttons and added bet input
- clean up style.css
- added getCurrency function to be built upon
- fixed issue with start new round clear player total text
- styling and DOM updates
- fixed comparative logic error
- updated styling of player details container
- updated player details container
- updated stand function to compare dealer vs player before dealerHit
- updated dealerHit functionality and checkDealerTotal log, added function to render dealer images face up after stand is hit
- updated commentary, created dealerHit function, and refactored playerHit
- updating notes for readability
- index and css updates, update hit and stand functionality & logic
- commiting changes to creatDeck index.html play and styles
- update backcard img
- cleaned up styling
- updated pseudocode and notes for future logic
- updated existing functions, added refreshdeck and cleardeck
- updated button styling
- removed Queue class, updated getplayercard getdealercard and added droptopcard
- added backcard image
- cleaned up play.js
- styled text for dealer and player
- added dealer and player text to respective containers
- updated card container styling
- added dynamic image creation to respective functions
- added image for backdeck card
- updated return from getCardImageName
- updating new div styling and buttons
- adding card container and divs for dealer, the deck and the player
- created flex div for buttons
- created dummy deal functions, tested card image rendering
- added hit, stand and start new round buttons
- updating image sizes
- updated body styling
- removed originally created DIV elements, created deal button element for future event listener
- moved deck creation into play.js, created a playQueue, and can now log the drawn or popped/dequeued cards image path
- created weight dictionary, shuffledeck and shufflearray function, and queue class
- adding pseudocode file and starting to list required functionality
- cleaning up js file
- created Card and Deck classes, Deck class has a createDeck function which is called and converted from an object to an array of cards
- updating some intial styling
- first commit, initial js, css & html files w/ some basic styling and card images
- Initial commit


src/play.ts
//*** VARIABLES ***///
var deck = new Deck();
deck.shuffleDeck();
// let cardSound = new Audio("assets/Audio/240777__f4ngy__dealing-card.wav");

var isDealClicked = false;
var canPlayerHit = false;
var canPlayerStand = false;
var hasBet = false;
var canStartNewRound = false;
var currencyTotal = "1000";
var winnings = "none";

let playerArray: (number | string | boolean )[] = [];
let dealerArray: (number | string | boolean )[] = [];

//*** HTML SELECTOR CONSTANTS ***///
const dealerCards = document.querySelector(".dealer-cards") as HTMLElement;
const playerCards = document.querySelector(".player-cards") as HTMLElement;
const playerTotal = document.querySelector(".player-total-text") as HTMLElement;
const outcomeText = document.querySelector(".outcome-text") as HTMLElement;
const currency = document.querySelector(".player-currency-text") as HTMLElement;
const playerBet =  document.querySelector(".player-bet-text") as HTMLElement;
const placeBetText = document.getElementById("submit-alert") as HTMLElement;
const alertPlayer = document.getElementById("zero-alert") as HTMLElement;

//*** BET LOGIC & PLAYER CURRENCY FUNCTIONALITY ***//
function getCurrency(){
    currency.innerText = currencyTotal;
    canStartNewRound = false;
}

function checkForZero(){
    if(currency.innerText == "0"){
        alertPlayer.innerText = "You have lost all your money! Ask the dealer for more by refreshing the page."
        outcomeText.innerText = "Dealer wins! You lost your bet."
        canStartNewRound = false;
    }
}

function placeBet() {
        if(hasBet == false){
            var betEle = document.getElementById("bet") as HTMLInputElement;
            var bet = parseInt(betEle.value)
            let newTotal = parseInt(currencyTotal) - bet;
            if(newTotal < 0){
                alertPlayer.innerText = "You can't bet more money than you have!";
                hasBet = false;
            } else {
                playerBet.innerHTML = bet.toString();
                placeBetText.innerText = "";
                alertPlayer.innerText = "";
                hasBet = true;
                canStartNewRound = false;
                currency.innerText = (parseInt(currencyTotal) - bet).toString();
                currencyTotal = currency.innerText;
            }
        }
    return playerBet.innerHTML;
}

function updateCurrency() {
    var betNum = parseInt(placeBet());
    var currencyTotalNum = parseInt(currencyTotal);
    if(winnings == "1"){
        currency.innerText = (currencyTotalNum + (betNum*2)).toString()
    } else if (winnings == "1.5"){
        currency.innerText = (currencyTotalNum + (betNum*2.5)).toString()
    } else if (winnings == "lose"){
        currency.innerText = currencyTotalNum.toString();
    } else if (winnings == "tie") {
        currency.innerText = (currencyTotalNum + betNum).toString();
    }
    betNum = 0;
    winnings = "none";
    currencyTotal = currency.innerText;
    canStartNewRound = true;
    checkForZero();
}


//*** PLAYER ***//

//create image element associated with first card off top of deck
function dealPlayerCard(){
    // cardSound.play();
    var createImage = document.createElement("img");
    createImage.src = getPlayerCard(drawTopCard());
    createImage.alt = "A picture of a playing card."
    playerCards.appendChild(createImage);
}

//updating players total weighted value of cards in hand to the DOM
function updatePlayerTotal(){
    playerTotal.innerHTML = updateTotal(playerArray)
}

//allows the player to hit (get) an additional card if necessary and checks players total weighted value
function playerHit(){
    if(canPlayerHit == true){
        dealPlayerCard();
        updatePlayerTotal();
        checkPlayerTotal();
    }
}

//*** DEALER ***//

//create image element for dealer associated with first card off top of deck (one of the two cards called in dealCards will be dealt face down)
function dealDealerCard(cardIsFaceUp = true){
    // cardSound.play();
    var createImage = document.createElement("img");
    var card = drawTopCard();
    card.cardIsFaceUp = cardIsFaceUp;
    createImage.src = getDealerCard(card);
    dealerCards.appendChild(createImage);
}

//render all card images face up after player has hit stand
function showDealerCard(){
    dealerCards.innerHTML = "";
    for (let i=0; i < dealerArray.length; i++){
        var createImage = document.createElement("img");
        createImage.src = `assets/${dealerArray[i].value}_of_${dealerArray[i].suit}.png`;
        dealerCards.appendChild(createImage);
    }
}

//dealerhit function to be performed after player has hit stand
function dealerHit(){
    if(updateTotal(dealerArray) < 17){
            do {
                dealDealerCard(true);
                updateTotal(dealerArray);
            } while (updateTotal(dealerArray) < 17);
    }
    updateTotal(dealerArray);
    checkDealerTotal();
}

//player can stand (stop hitting) at which time the dealer will hit or show their cards to end the round
function stand(){
    if(canPlayerStand == true){
    canPlayerHit = false;
    showDealerCard();
    checkDealerTotal();
    }
}

//check total weighted value of cards in hand
function updateTotal(array: (any)[]){
    let sum = 0;
    for(let i=0; i < array.length; i++){
        // console.log(array[i])
        sum += array[i].weight;
    }
    return sum;
}

//*** COMPARATIVE LOGIC & VALUE CHECKING ***//
function checkPlayerTotal(){
    if (updateTotal(playerArray) > 21)
    {
        showDealerCard();
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
        updateCurrency();
    } else if (updateTotal(playerArray) == 21) {
        showDealerCard();
        if(updateTotal(playerArray) == updateTotal(dealerArray)){
            outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
            winnings = "tie";
            updateCurrency();
        } else if (updateTotal(playerArray) == 21 && updateTotal(dealerArray) < 17 ) {
            dealerHit();
            canPlayerHit = false;
        } else if (updateTotal(playerArray) == 21 && updateTotal(dealerArray) >= 17 ) {
            outcomeText.innerText = "You have won your bet! Click start new round.";
            winnings = "1";
            updateCurrency();
        }
        canPlayerHit = false;
        canStartNewRound = true;
        canPlayerStand = false;
    }
}

function checkDealerTotal(){
    if(updateTotal(dealerArray) < 17){
        dealerHit();
    }else if (updateTotal(dealerArray) > 21)
    {
        outcomeText.innerText = "You have won your bet! Click start new round.";
        winnings = "1";
    } else if (updateTotal(dealerArray) == 21 && updateTotal(playerArray) !== 21){
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(dealerArray) > 17 && updateTotal(dealerArray) > updateTotal(playerArray)){
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(dealerArray) == 21 && updateTotal(playerArray) == 21){
        outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
        winnings = "tie";
    } else {
        compareTotals();
    }
    canStartNewRound = true;
    canPlayerStand = false;
    updateCurrency();
}

function compareTotals(){
    if(updateTotal(dealerArray) > 21){
        checkDealerTotal();
    } else if (updateTotal(playerArray) > updateTotal(dealerArray)){
        outcomeText.innerText = "You have won your bet! Click start new round.";
        winnings = "1";
    } else if (updateTotal(playerArray) < updateTotal(dealerArray)) {
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(playerArray) == updateTotal(dealerArray)) {
        outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
        winnings = "tie";
    }
    canStartNewRound = true;
    canPlayerStand = false;
    updateCurrency();
}


//function to create new deck and shuffle again when new round is started
function refreshDeck(){
    deck = new Deck();
    deck.shuffleDeck();
}

//deal intial cards to player and dealer
var dealCards = function(){
    if(isDealClicked == false) { 
        this.removeEventListener("click", dealCards);
    }
    
    if (hasBet == true){
        setTimeout(function() {
            dealPlayerCard();
        }, 500);

        setTimeout(function() {
            dealDealerCard(false);
        }, 1000);

        setTimeout(function() {
            dealPlayerCard();
        }, 1500);

        setTimeout(function() {
            dealDealerCard();
        }, 2000);

        setTimeout(function() {
            updatePlayerTotal();
            
    
            if(updateTotal(playerArray) == 21) {
                showDealerCard();
                outcomeText.innerText = "BLACKJACK! You win 1.5 times your bet. Click start new round.";
                canPlayerHit = false;
                canPlayerStand = false;
                canStartNewRound = true;
                winnings = "1.5";
                updateCurrency();
            }
            checkPlayerTotal();
            canPlayerHit = true;
            canPlayerStand = true;
            placeBetText.innerText = "";
        }, 2000);
    }
}

//clear table - remove card images, player total, hand outcome, refresh the deck and empty dealer and player array
function clearTable(){
    if(canStartNewRound == true) {
        dealerCards.innerHTML = "";
        playerCards.innerHTML = "";
        playerTotal.innerText = "";
        outcomeText.innerText = "";
        playerBet.innerHTML = "";
        refreshDeck();
        document.getElementById("betform").reset();
        playerArray = [];
        dealerArray = [];
        canPlayerStand = false;
        canPlayerHit = false;
        hasBet = false;
        placeBetText.innerText = "Submit your bet below & hit Deal above!";
    }
}

//*** HANDLERS ***//
document.querySelector('.deal').addEventListener("click", dealCards);
document.querySelector('.hit').addEventListener("click", playerHit);
document.querySelector('.stand').addEventListener("click", stand);
document.querySelector('.newround').addEventListener("click", function(){
    document.querySelector('.deal').addEventListener("click", dealCards);
    clearTable();
})
document.querySelector('#bet-submit').addEventListener("click", placeBet);


//*** ON LOAD ***//

window.onload = function() {
    getCurrency();
}

createDeck.js
//set up card value array, suit array, CARDWEIGHT dictionary and create deck
let cardValue = ['2','3','4','5','6','7','8','9','10','jack','queen','king','ace'];

let cardSuit = ['clubs','diamonds','hearts','spades'];

//dictionary used to add weighted values to the face cards and ace
//source: https://pietschsoft.com/post/2015/09/05/javascript-basics-how-to-create-a-dictionary-with-keyvalue-pairs
const CARDWEIGHT = {
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "10": 10,
    "jack": 10,
    "queen": 10,
    "king": 10,
    "ace": 11,
}

class Card {
    constructor(value, suit) {
        this.value = value;
        this.suit = suit;
        this.weight = CARDWEIGHT[value];
        this.cardIsFaceUp = false;
    }

}

class Deck {
    constructor(cards = createDeck()) {
        this.cards = cards;
    }

    shuffleDeck() {
        this.cards = shuffleArray(this.cards);
    }

}

//creating the deck of cards using flatMap, using just map gave back and array with a 4 indices and 13 cards in each
//flatmap: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap
function createDeck() {
    return cardValue.flatMap(value => {
        return cardSuit.map(suit =>{
            return new Card(value,suit)
        })
    })
}


//shuffles the created deck - grabs a random index, and swaps it with the current index over the length of the array
// source: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm
function shuffleArray(array) {
    for (var i = array.length - 1; i > 0; i--) {

        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));

        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    return array;
}

//grabs the next card for the players hand
function getPlayerCard(card){
    playerArray.push(card)
    return `assets/${card.value}_of_${card.suit}.png`
}

//grabs the next card for the dealers hand.. if cardIsFaceUp is passed as false that card is dealt face down
function getDealerCard(card){
    dealerArray.push(card)
    if (card.cardIsFaceUp) {
        return `assets/${card.value}_of_${card.suit}.png`
    }
    return `assets/backcard.png`
}

//draws the first card from the shuffled deck
function drawTopCard(){
    return deck.cards.shift();
}

src/createDeck.ts
//set up card value array, suit array, CARDWEIGHT dictionary and create deck
let cardValue = ['2','3','4','5','6','7','8','9','10','jack','queen','king','ace'];

let cardSuit = ['clubs','diamonds','hearts','spades'];

//dictionary used to add weighted values to the face cards and ace
//source: https://pietschsoft.com/post/2015/09/05/javascript-basics-how-to-create-a-dictionary-with-keyvalue-pairs
const CARDWEIGHT = {
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "10": 10,
    "jack": 10,
    "queen": 10,
    "king": 10,
    "ace": 11,
}

class Card {
    value: number;
    suit: string;
    weight: number;
    cardIsFaceUp: boolean;
    
    constructor(value, suit) {
        this.value = value;
        this.suit = suit;
        this.weight = CARDWEIGHT[value];
        this.cardIsFaceUp = false;
    }

}

class Deck {
    constructor(cards = createDeck()) {
        this.cards = cards;
    }

    shuffleDeck() {
        this.cards = shuffleArray(this.cards);
    }

}

//creating the deck of cards using flatMap, using just map gave back and array with 4 indices and 13 cards in each
//flatmap: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap
function createDeck() {
    return cardValue.flatMap(value => {
        return cardSuit.map(suit =>{
            return new Card(value,suit)
        })
    })
}


//shuffles the created deck - grabs a random index, and swaps it with the current index over the length of the array
// source: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm
function shuffleArray(array) {
    for (var i = array.length - 1; i > 0; i--) {

        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));

        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    return array;
}

//grabs the next card for the players hand
function getPlayerCard(card){
    playerArray.push(card)
    return `assets/${card.value}_of_${card.suit}.png`
}

//grabs the next card for the dealers hand.. if cardIsFaceUp is passed as false that card is dealt face down
function getDealerCard(card){
    dealerArray.push(card)
    if (card.cardIsFaceUp) {
        return `assets/${card.value}_of_${card.suit}.png`
    }
    return `assets/backcard.png`
}

//draws the first card from the shuffled deck
function drawTopCard(){
    return deck.cards.shift();
}

play.js
//*** VARIABLES ***///
var deck = new Deck();
deck.shuffleDeck();
// let cardSound = new Audio("assets/Audio/240777__f4ngy__dealing-card.wav");

var isDealClicked = false;
var canPlayerHit = false;
var canPlayerStand = false;
var hasBet = false;
var canStartNewRound = false;
var currencyTotal = "1000";
var winnings = "none";

let playerArray = [];
let dealerArray = [];

//*** HTML SELECTOR CONSTANTS ***///
const dealerCards = document.querySelector(".dealer-cards");
const playerCards = document.querySelector(".player-cards");
const playerTotal = document.querySelector(".player-total-text");
const outcomeText = document.querySelector(".outcome-text");
const currency = document.querySelector(".player-currency-text");
const playerBet =  document.querySelector(".player-bet-text");
const placeBetText = document.getElementById("alert");
const alertPlayer = document.getElementById("zero-alert");

//*** BET LOGIC & PLAYER CURRENCY FUNCTIONALITY ***//
function getCurrency(){
    currency.innerText = currencyTotal;
    canStartNewRound = false;
}

function checkForZero(){
    if(currency.innerText == "0"){
        alertPlayer.innerText = "You have lost all your money! Ask the dealer for more by refreshing the page."
        outcomeText.innerText = "Dealer wins! You lost your bet."
        canStartNewRound = false;
    }
}

function placeBet() {
        if(hasBet == false){
            var bet = document.getElementById("bet").value;
            let newTotal = parseInt(currencyTotal) - bet;
            if(newTotal < 0){
                alertPlayer.innerText = "You can't bet more money than you have!";
                hasBet = false;
            } else {
                playerBet.innerHTML = bet;
                placeBetText.innerText = "";
                alertPlayer.innerText = "";
                hasBet = true;
                canStartNewRound = false;
                currency.innerText = parseInt(currencyTotal) - bet;
                currencyTotal = currency.innerText;
            }
        }
    return playerBet.innerHTML;
}

function updateCurrency() {
    var betNum = parseInt(placeBet());
    var currencyTotalNum = parseInt(currencyTotal);
    if(winnings == "1"){
        currency.innerText = (currencyTotalNum + (betNum*2))
    } else if (winnings == "1.5"){
        currency.innerText = (currencyTotalNum + (betNum*2.5))
    } else if (winnings == "lose"){
        currency.innerText = currencyTotalNum;
    } else if (winnings == "tie") {
        currency.innerText = currencyTotalNum + betNum;
    }
    betNum = 0;
    winnings = "none";
    currencyTotal = parseInt(currency.innerText);
    canStartNewRound = true;
    checkForZero();
}


//*** PLAYER ***//

//create image element associated with first card off top of deck
function dealPlayerCard(){
    // cardSound.play();
    var createImage = document.createElement("img");
    createImage.src = getPlayerCard(drawTopCard());
    createImage.alt = "A picture of a playing card."
    playerCards.appendChild(createImage);
}

//updating players total weighted value of cards in hand to the DOM
function updatePlayerTotal(){
    playerTotal.innerHTML = updateTotal(playerArray)
}

//allows the player to hit (get) an additional card if necessary and checks players total weighted value
function playerHit(){
    if(canPlayerHit == true){
        dealPlayerCard();
        updatePlayerTotal();
        checkPlayerTotal();
    }
}

//*** DEALER ***//

//create image element for dealer associated with first card off top of deck (one of the two cards called in dealCards will be dealt face down)
function dealDealerCard(cardIsFaceUp = true){
    // cardSound.play();
    var createImage = document.createElement("img");
    var card = drawTopCard();
    card.cardIsFaceUp = cardIsFaceUp;
    createImage.src = getDealerCard(card);
    dealerCards.appendChild(createImage);
}

//render all card images face up after player has hit stand
function showDealerCard(){
    dealerCards.innerHTML = "";
    for (let i=0; i < dealerArray.length; i++){
        var createImage = document.createElement("img");
        createImage.src = `assets/${dealerArray[i].value}_of_${dealerArray[i].suit}.png`;
        dealerCards.appendChild(createImage);
    }
}

//dealerhit function to be performed after player has hit stand
function dealerHit(){
    if(updateTotal(dealerArray) < 17){
            do {
                dealDealerCard(true);
                updateTotal(dealerArray);
            } while (updateTotal(dealerArray) < 17);
    }
    updateTotal(dealerArray);
    checkDealerTotal();
}

//player can stand (stop hitting) at which time the dealer will hit or show their cards to end the round
function stand(){
    if(canPlayerStand == true){
    canPlayerHit = false;
    showDealerCard();
    checkDealerTotal();
    }
}

//check total weighted value of cards in hand
function updateTotal(array){
    let sum = 0;
    for(let i=0; i < array.length; i++){
        // console.log(array[i])
        sum += array[i].weight;
    }
    return sum;
}

//*** COMPARATIVE LOGIC & VALUE CHECKING ***//
function checkPlayerTotal(){
    if (updateTotal(playerArray) > 21)
    {
        showDealerCard();
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
        updateCurrency();
    } else if (updateTotal(playerArray) == 21) {
        showDealerCard();
        if(updateTotal(playerArray) == updateTotal(dealerArray)){
            outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
            winnings = "tie";
            updateCurrency();
        } else if (updateTotal(playerArray) == 21 && updateTotal(dealerArray) < 17 ) {
            dealerHit();
            canPlayerHit = false;
        } else if (updateTotal(playerArray) == 21 && updateTotal(dealerArray) >= 17 ) {
            outcomeText.innerText = "You have won your bet! Click start new round.";
            winnings = "1";
            updateCurrency();
        }
        canPlayerHit = false;
        canStartNewRound = true;
        canPlayerStand = false;
    }
}

function checkDealerTotal(){
    if(updateTotal(dealerArray) < 17){
        dealerHit();
    }else if (updateTotal(dealerArray) > 21)
    {
        outcomeText.innerText = "You have won your bet! Click start new round.";
        winnings = "1";
    } else if (updateTotal(dealerArray) == 21 && updateTotal(playerArray) !== 21){
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(dealerArray) > 17 && updateTotal(dealerArray) > updateTotal(playerArray)){
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(dealerArray) == 21 && updateTotal(playerArray) == 21){
        outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
        winnings = "tie";
    } else {
        compareTotals();
    }
    canStartNewRound = true;
    canPlayerStand = false;
    updateCurrency();
}

function compareTotals(){
    if(updateTotal(dealerArray) > 21){
        checkDealerTotal();
    } else if (updateTotal(playerArray) > updateTotal(dealerArray)){
        outcomeText.innerText = "You have won your bet! Click start new round.";
        winnings = "1";
    } else if (updateTotal(playerArray) < updateTotal(dealerArray)) {
        outcomeText.innerText = "Dealer wins! You lost your bet. Click start new round.";
        winnings = "lose";
    } else if (updateTotal(playerArray) == updateTotal(dealerArray)) {
        outcomeText.innerText = "It's a tie! You get your bet back. Click start new round.";
        winnings = "tie";
    }
    canStartNewRound = true;
    canPlayerStand = false;
    updateCurrency();
}


//function to create new deck and shuffle again when new round is started
function refreshDeck(){
    deck = new Deck();
    deck.shuffleDeck();
}

//deal intial cards to player and dealer
var dealCards = function(){
    if(isDealClicked == false) { 
        this.removeEventListener("click", dealCards);
    }
    
    if (hasBet == true){
        setTimeout(function() {
            dealPlayerCard();
        }, 500);

        setTimeout(function() {
            dealDealerCard(false);
        }, 1000);

        setTimeout(function() {
            dealPlayerCard();
        }, 1500);

        setTimeout(function() {
            dealDealerCard();
        }, 2000);

        setTimeout(function() {
            updatePlayerTotal();
            
    
            if(updateTotal(playerArray) == 21) {
                showDealerCard();
                outcomeText.innerText = "BLACKJACK! You win 1.5 times your bet. Click start new round.";
                canPlayerHit = false;
                canPlayerStand = false;
                canStartNewRound = true;
                winnings = "1.5";
                updateCurrency();
            }
            checkPlayerTotal();
            canPlayerHit = true;
            canPlayerStand = true;
            placeBetText.innerText = "";
        }, 2000);
    }
}

//clear table - remove card images, player total, hand outcome, refresh the deck and empty dealer and player array
function clearTable(){
    if(canStartNewRound == true) {
        dealerCards.innerHTML = "";
        playerCards.innerHTML = "";
        playerTotal.innerText = "";
        outcomeText.innerText = "";
        playerBet.innerHTML = "";
        refreshDeck();
        document.getElementById("betform").reset();
        playerArray = [];
        dealerArray = [];
        canPlayerStand = false;
        canPlayerHit = false;
        hasBet = false;
        placeBetText.innerText = "Submit your bet below & hit Deal above!";
    }
}

//*** HANDLERS ***//
document.querySelector('.deal').addEventListener("click", dealCards);
document.querySelector('.hit').addEventListener("click", playerHit);
document.querySelector('.stand').addEventListener("click", stand);
document.querySelector('.newround').addEventListener("click", function(){
    document.querySelector('.deal').addEventListener("click", dealCards);
    clearTable();
})
document.querySelector('#bet-submit').addEventListener("click", placeBet);


//*** ON LOAD ***//

window.onload = function() {
    getCurrency();
}